pkg/build/prometheus/metrics.go

```diff
@@ -0,0 +1,109 @@
+package prometheus
+
+// NOTE, this module explicitly attempts to avoid references to the build api types/structs
+// in an attempt to start breaking up our compile dependencies
+
+import (
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// BuildSubsystem is how this build controller is represented in
+// prometheus metrics.
+const BuildSubsystem = "build_controller"
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	buildsCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_counters",
+			Help:      "Counts total teriminal builds by phase, and in case of failures, reason",
+		},
+		[]string{"phase", "reason"},
+	)
+	buildsGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_gauges",
+			Help:      "Show current counts of non-terminal builds by phase",
+		},
+		[]string{"phase"},
+	)
+	buildsOverallHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_timing_histograms",
+			Help:      "Histogram of durations of builds in seconds that have exited by phase",
+		},
+		[]string{"phase"},
+	)
+	buildStagesHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_stage_timing_histograms",
+			Help:      "Histogram of durations of stages in milliseconds that have exited by stage name",
+		},
+		[]string{"stage"},
+	)
+	buildStageStepsHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: BuildSubsystem,
+			Name:      "build_stage_step_timing_histograms",
+			Help:      "Histogram of durations of steps in milliseconds that have exited by stage name and step name",
```

Comment:
"Histogram of durations in milliseconds of steps that have finished, by stage name and step name